### ISSUE TITLE
more combinators

### ISSUE THREAD

**Issue by davidchambers** [sanctuary-js/sanctuary#139]
@Avaq raised the issue of combinators on Gitter. Sanctuary currently exports [`I`][1] and [`K`][2], but [fantasy-combinators][3] defines several more, neatly summarized in a [gist][4]:

> ```javascript
> const I = x => x;
> const K = x => _ => x;
> const A = f => x => f(x);
> const T = x => f => f(x);
> const C = f => y => x => f(x)(y);
> const B = f => g => x => f(g(x));
> const S = f => g => x => f(x)(g(x));
> const P = f => g => x => y => f(g(x))(g(y));
> const Y = f => (g => g(g))(g => f(x => g(g)(x)));
> ```

`A` is a strict version of [`R.call`][5] (since it expects a unary function).

`C` is a strict version of [`S.flip`][6] (since it expects a curried function).

`B` is [`S.compose`][7].

I'd like Sanctuary to provide more of these combinators. I'd welcome a pull request for one or more of these functions. :)

[1]: https://github.com/plaid/sanctuary#I
[2]: https://github.com/plaid/sanctuary#K
[3]: https://github.com/fantasyland/fantasy-combinators
[4]: https://gist.github.com/Avaq/1f0636ec5c8d6aed2e45#file-combinators-js
[5]: http://ramdajs.com/docs/#call
[6]: https://github.com/plaid/sanctuary#flip
[7]: https://github.com/plaid/sanctuary#compose

**Comment by Avaq** [comment]
Hi @davidchambers!

In all honesty I have to say I'm still a bit skeptical about the decision to add *more combinators* to Sanctuary. There are several points I'd like to make.

----

Firstly, as I've come to understand, combinators have their roots in Lambda calculus. The Y combinator specifically -I believe- is a way to gain recursion in an environment where there is no variable assignment other than Lambda binding. That's the main problem is solves. The point I'm trying to make here is that we *can* do other types of variable assignment in JavaScript, so it's really quite easy to create recursive functions without help from the Y combinator:

```js
//With Y:
const factorial = Y(f => n => n < 1 ? 1 : n * f(n - 1));
//Easier to read and just as easy to write without Y:
const factorial = n => n < 1 ? 1 : n * factorial(n - 1);
```

So the only application it has in JavaScript (as far as I can see), is to enable writing point-free recursive functions:

```js
//Without Y:
const helper = n => factorial(n); //I need this pointful helper :(
const factorial = ifElse(isZero, K(1), S(mult)(B(helper)(dec)));
//With Y:
const factorial = Y(B(ifElse(isZero, K(1)))(B(S(mult))(C(B)(dec))));
```

Now, besides the fact that it took me close to two hours to create the point-free version ([marvel at its beauty](http://goo.gl/p31h8l)), it's also quite cryptic as far as readable code goes. This leads me to conclude that usage of the Y-combinator is very practical in lambda-only environments, but maybe not so much in environments that support self-referential functions.

----

Secondly, I think the idea in combinatory-logic is for the combinators to be defined in terms of only unary lambda's or other combinators. Defining them using `sanctuary-def` defeats that philosophy/practice, which is okay because we're basically just porting over the utility that they provide. But then wouldn't it be better to just name the utility functions after the problem they intend to solve? And wouldn't you then end up with `S.flip` and `S.compose` etc?

----

Thirdly, it's still my opinion that Sanctuary is better off remaining lean, and including only the necessary functions to allow for safe JavaScript. It already has a bunch of utility functions which could be classified as combinators, which it includes out of need.

----

Conclusion: I said I would gladly submit a pull-request for the Y combinator, and as much as I would like to find out if `sanctuary-def` can deal with its absurd type signature, I can't help but wonder how much value I would really be adding to the code-base.

I'd love to hear your thoughts!

**Comment by Avaq** [comment]
I would probably be happy with `S` being added (possibly with an alias `substitute`), `C` being added as an alias to `flip` and `B` as an alias to `compose`. This strikes me as a nice balance between brevity, utility and accessibility. Then again, I'm not sure what you're aiming for. :)

**Comment by davidchambers** [comment]
> I would probably be happy with `S` being added (possibly with an alias `substitute`), `C` being added as an alias to `flip` and `B` as an alias to `compose`. This strikes me as a nice balance between brevity, utility and accessibility.

This sounds great to me. I think `A` would be useful as well. I'm surprised by how often I see comments on Gitter from people trying `R.call(R.__, arg)`, which doesn't work since [`R.call`][1] supports variadic functions. In this case `T` would be even more convenient than `A` + placeholder, so perhaps we should provide `T` as well.

Your comments about the impracticality of the Y combinator in JavaScript are well taken, @Avaq. What most excited me about your recent comments on Gitter was seeing a type signature for what has always seemed to me a mysterious function. Including the function in Sanctuary would be valuable if it improves our understanding of it, even if we never actually use it in a program. The cost is relatively low.

Let's not include `Y` at this stage, but if you're interested in contributing a pull request for one or more of the other functions, please do so. :)


[1]: http://ramdajs.com/docs/#call

**Comment by whatsajava** [comment]
hey good project lol

**Comment by whatsajava** [comment]
hey bad project lol

**Comment by Avaq** [comment]
:information_source: I'm working on a PR for this.

**Comment by whatsajava** [comment]
metoohaha


________________________________

**Comment by Avaq** [comment]
@whatsajava Oh wow. Haha! I just finished mine. What do we do?

**Comment by jethrolarson** [comment]
Aw yes, the [`C, A, T, S, P, Y, B, I, K` system](http://www.unp.me/attachments/f98/11203d1297789674-invisible-bike-invisible-bike.jpg)

**Comment by davidchambers** [comment]
I just looked at some code and spotted multiple occurrences of `R.lift(R.call)(f, S.I)`. Here's a concrete example:

```javascript
lift(call)(pipe([prop('category'),
                 categories.fromHierarchy,
                 map(get(String, 'group')),
                 map(or),
                 map(over(L.group)),
                 fromMaybe(I)]),
           I)
```

The pipeline takes a value of some type and returns a function which operates on a value of that type. This function is then applied to the original value. I wrote a function to obviate the need for `lift(call)(…, I)` shenanigans, and gave it a type signature: `(a -> a -> b) -> a -> b`. I checked the [gist][1] and discovered that this is the W combinator!

I'll define `W` in my own code, as I want to start using it right away, but we should add it to Sanctuary. :)


[1]: https://gist.github.com/Avaq/1f0636ec5c8d6aed2e45#file-combinators-md

**Comment by davidchambers** [comment]
I just merged #221, which added `T`, so we now have all the combinators listed in the issue description other than `P` and `Y`. We may decide to add even more combinators at some point, but our current suite feels sufficient to close this issue. :)
